### PR TITLE
Add Access List usage events, emit event for userloginstate Generator.

### PR DESCRIPTION
**Pull request by mdwn**
Access List usage events have been added, counting create and delete for Access Lists and Access List members. Additionally, an event for granting Access List access to users has been added to the userloginstate Generator and is now tested for.

Note: I haven't actually tested this yet, I'm still trying to figure out how to do this, or if it's possible.

### CODE REVIEW

**Comment by mdwn** [comment]
It looks like I've missed a few things w/r/t prehog here, so I need to add these in. I'll ping once these are finished.

File: lib/auth/userloginstate/generator.go [hunk 1]
```diff
@@ -120,6 +137,11 @@ func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginst
 		return nil, trace.Wrap(err)
 	}
 
+	// Emit the usage event metadata.
+	if err := g.emitUsageEvent(ctx, user, uls); err != nil {
+		g.log.WithError(err).Debug("Error emitting usage event during user login state generation")
```
Comment (zmb3) [inline lib/auth/userloginstate/generator.go:142]: `WithError` will dump a stack trace which implies that you want to draw attention to this in the logs, but debug level implies that it's not critical. I would either remove the `WithError` or make the level warn.

File: lib/auth/userloginstate/generator.go [hunk 2]
```diff
@@ -59,6 +70,10 @@ func (g *GeneratorConfig) CheckAndSetDefaults() error {
 		return trace.BadParameter("missing access")
 	}
 
+	if g.UsageEvents == nil {
+		return trace.BadParameter("missing usage events")
```
[the remark this answers is not in the document]
Comment (mdwn) [reply lib/auth/userloginstate/generator.go:74]: I suppose it's not required, but the Usage Reporter looks like it's always enabled regardless of whether this is OSS/Cloud/Enterprise. However, looking around OSS it looks like there are other examples of checking for Cloud before emitting, so I'll make this optional based on that.

File: lib/auth/userloginstate/generator.go [hunk 3]
```diff
@@ -120,6 +140,13 @@ func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginst
 		return nil, trace.Wrap(err)
 	}
 
+	if modules.GetModules().Features().Cloud {
```
Comment (zmb3) [inline lib/auth/userloginstate/generator.go:143]: Maybe do the cloud check (which accesses a global under lock) only in the check and set defaults, and from there on just check if your `usageEvents` client is not nil?

File: proto/prehog/v1alpha/teleport.proto [hunk 4]
```diff
@@ -701,6 +701,67 @@ message AssistActionEvent {
...
+message AccessListMemberDeleteEvent {
+  // user_name is anonymized user name
+  string user_name = 1;
+  AccessListMetadata metadata = 2;
+}
+
+// AccessListGrantsToUser is an event that is emitted when access list permissions are granted to a user
+// on login.
+message AccessListGrantsToUserEvent {
+  // UserName is anonymized user name
```
Comment (avatus) [inline proto/prehog/v1alpha/teleport.proto:755]:
```suggestion
  // user_name is the anonymized user name
```
or maybe "an" idk

**Review by avatus — APPROVED** [review]

**Review by Joerger — APPROVED** [review]

**Comment by espadolini** [comment]
There's not even a PR up for this in cloud, so we just narrowly avoided a conflict in the protobuf definition, and the events generated here are just going to end up as errors in the ingest service's logs, because they're just parsed as unknown fields in the message. We should revert the v13 and v14 backports until we fix that.

**Comment by espadolini** [comment]
(see https://github.com/gravitational/cloud/blob/master/rfd/0070-usage-reporting-dev-guidelines.md#teleport )

**Pull request merged**